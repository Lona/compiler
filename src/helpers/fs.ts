import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
