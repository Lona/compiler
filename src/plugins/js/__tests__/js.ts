import path from 'path'
import fs from 'fs'
import { createFs, toJSON, copy } from 'buffs'
import plugin from '../index'
import Helpers from '../../../helpers'

const workspacePath = path.join(__dirname, '../../../../examples/workspace')

const tokensBlock = (string: string) => '```tokens\n' + string + '\n```\n'

// TODO: Test more kinds of JS
describe('JS', () => {
  it('generates js', async () => {
    const source = createFs({
      'lona.json': JSON.stringify({}),
      'Colors.md': tokensBlock(`let color: Color = #color(css: "pink")`),
    })

    const helpers = Helpers(source, '/')

    await plugin.convertWorkspace('/', helpers, { output: '/output' })

    const index = source.readFileSync('/output/index.js', 'utf8')
    const colors = source.readFileSync('/output/Colors.js', 'utf8')

    expect(index).toMatchSnapshot()
    expect(colors).toMatchSnapshot()
  })

  it('converts workspace', async () => {
    const source = createFs()

    copy(fs, source, workspacePath, '/')

    const helpers = Helpers(source, '/')

    await plugin.convertWorkspace('/', helpers, { output: '/output' })

    expect(toJSON(source, '/output')).toMatchSnapshot()
  })
})
