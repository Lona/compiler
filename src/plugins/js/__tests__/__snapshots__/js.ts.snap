// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS converts workspace 1`] = `
Object {
  "/output/Colors.js": "var primary = \\"#45CBFF\\";
module.exports.primary = primary;

var accent = primary;
module.exports.accent = accent;

var testSaturate = color.saturate(accent, 0.3);
module.exports.testSaturate = testSaturate;
",
  "/output/Shadows.js": "var small = shadow(0, 2, 2, 0, primary);
module.exports.small = small;
",
  "/output/TextStyles.js": "var heading1 = textStyle(optional.value(\\"Helvetica\\"), optional.value(28), fontWeight.w700, optional.value(\\"teal\\"));
module.exports.heading1 = heading1;
",
  "/output/index.js": "var __lona_import_0 = require(\\"./Colors\\");
Object.keys(__lona_import_0).forEach(function (key) {
  Object.defineProperty(module.exports, key, {
    enumerable: true,
    get: function get() {
      return __lona_import_0[key];
    }
  });
})

var __lona_import_1 = require(\\"./Shadows\\");
Object.keys(__lona_import_1).forEach(function (key) {
  Object.defineProperty(module.exports, key, {
    enumerable: true,
    get: function get() {
      return __lona_import_1[key];
    }
  });
})

var __lona_import_2 = require(\\"./TextStyles\\");
Object.keys(__lona_import_2).forEach(function (key) {
  Object.defineProperty(module.exports, key, {
    enumerable: true,
    get: function get() {
      return __lona_import_2[key];
    }
  });
})",
}
`;

exports[`JS generates js 1`] = `
"var __lona_import_0 = require(\\"./Colors\\");
Object.keys(__lona_import_0).forEach(function (key) {
  Object.defineProperty(module.exports, key, {
    enumerable: true,
    get: function get() {
      return __lona_import_0[key];
    }
  });
})"
`;

exports[`JS generates js 2`] = `
"var color = \\"pink\\";
module.exports.color = color;
"
`;
